pkg/controllers/subnetset/subnetset_controller.go

```diff
@@ -261,39 +261,30 @@ func (r *SubnetSetReconciler) CollectGarbage(ctx context.Context) {
 	defer func() {
 		log.Info("SubnetSet garbage collection completed", "duration(ms)", time.Since(startTime).Milliseconds())
 	}()
-	subnetSetList := &v1alpha1.SubnetSetList{}
-	err := r.Client.List(ctx, subnetSetList)
+
+	crdSubnetSetIDs, err := r.listSubnetSetIDsFromCRs(ctx)
 	if err != nil {
-		log.Error(err, "Failed to list SubnetSet CR")
-		return
-	}
-	var nsxSubnetList []*model.VpcSubnet
-	for _, subnetSet := range subnetSetList.Items {
-		nsxSubnetList = append(nsxSubnetList, r.SubnetService.ListSubnetCreatedBySubnetSet(string(subnetSet.UID))...)
-	}
-	if len(nsxSubnetList) == 0 {
+		log.Error(err, "Failed to list Subnet CRs")
 		return
 	}
+	crdSubnetSetIDsSet := sets.NewString(crdSubnetSetIDs...)
 
-	subnetSetIDs := sets.New[string]()
-	for _, subnetSet := range subnetSetList.Items {
-		if err := r.deleteSubnetForSubnetSet(subnetSet, true); err != nil {
-			metrics.CounterInc(r.SubnetService.NSXConfig, metrics.ControllerDeleteFailTotal, MetricResTypeSubnetSet)
-		} else {
-			metrics.CounterInc(r.SubnetService.NSXConfig, metrics.ControllerDeleteSuccessTotal, MetricResTypeSubnetSet)
-		}
-		subnetSetIDs.Insert(string(subnetSet.UID))
-	}
```

Comment:
I notice 2 issues:

1. We need to keep this part. There are 2 steps in subnetset gc: 1) delete the subnets that belongs to a subnetset which does not have CR in k8s, which has already implemented in this PR, 2) for subnet under subnetset with CR in k8s, we also need to delete it if there is no subnetport under it, which is implemented in this deleted block.

2) When removing subnetset finalizer we slightly broke this function. If you take a look at the last version, where it defines `func (r *SubnetSetReconciler) DeleteSubnetForSubnetSet(obj v1alpha1.SubnetSet, updataStatus bool) (bool, error)` which does not return an error when `portNums > 0`. This is used for the case to skip deleting the subnet with subnetport.

Reply:
For the first one:
Not understand why  `for subnet under subnetset with CR in k8s, we also need to delete it if there is no subnetport under it`?        
    
For the second one:
In the current code, it will return an error when `portNums > 0`. it is in the `deleteSubnets` function, which is called by `deleteSubnetForSubnetSet`.
```
func (r *SubnetSetReconciler) deleteSubnets(nsxSubnets []*model.VpcSubnet) error {
	for _, nsxSubnet := range nsxSubnets {
		portNums := len(r.SubnetPortService.GetPortsOfSubnet(*nsxSubnet.Id))
		if portNums > 0 {
			return fmt.Errorf("fail to delete Subnet/%s from SubnetSet CR, there is stale ports", *nsxSubnet.Id)
		}
```

Reply:
Yes, @yanjunz97  is right, this delete the original fix from
https://github.com/vmware-tanzu/nsx-operator/pull/636
if there is portNums, you should continue instead of return, 
because there are multiple subnetsets under the same subnetset.

Also, you change lockSubnet order, which should be put before read portNum,
this is logic in @yanjunz97 PR:
https://github.com/vmware-tanzu/nsx-operator/pull/770/files#diff-ea8fdc4a9adcc30f1ee646f03f10d565526b3a2b6fec57ae9abaa316c14b8d89R296
Please change these logic back as before in DeleteSubnetForSubnetSet function.

Reply:
For the first one: @yanjunz97 is right
this was mentioned in PR:
https://github.com/vmware-tanzu/nsx-operator/pull/769/files#r1784012526

For the question:
`for subnet under subnetset with CR in k8s, we also need to delete it if there is no subnetport under it?
`
subnetset has this specific GC process, 
The normal GC process is to delete stale NSX subnet if their corresponding K8s subnet/subnset CR has been deleted.

The other GC is to delete NSX subnet scale down in the subnetset CR,
when the nsx subnet allocated from subnetset CR has no more ports, we should GC this nsx subnet and reclaim it, but this GC shoud be only for 
subnetset GC @dantingl  please correct me if misunderstanding. 
Previous Janwei's GC logic is to do this logic, but has some issue,s
we should fix this logic and add the normal GC process as you are doing in this PR.

Reply:
Updated.